### PR TITLE
Update LLVM to 11 and add LLVM utilities

### PR DESCRIPTION
**Pull request by sunflsks**
Sorry for the large PR

This adds the following:
- lld (the LLVM linker)
- clang-tools (clangd, clang-format, sancov, and a few other tools)
- llvm-utils (llc, lli, and other tools for interacting with bitcode and LLVM)
- libllvm-polly (a new polyhedral optimizer for LLVM)
- LLVM libiosexec support

The entire LLVM suite needs to be bumped up a packaging version, due to the fact that some of the extra projects add symbols to libLLVM.so that are not there by default.

### CODE REVIEW

**Comment by CRKatri** [comment]
Any chance we could get llvm-ar llvm-strip llvm-no etc.

**Comment by sunflsks** [comment]
This has all 3 of those

**Comment by CRKatri** [comment]
Oh, well you packaging needs to be split even more, iirc clang-tools-extra is split, a package for each bin.

**Comment by CRKatri** [comment]
All the control files need to be updated to 11, it needs to be even more debianized, and most importantly we need lldb, preferably a fat arm63{,e} one but even just arm64 would work.

**Comment by CRKatri** [comment]
swiftpm would also be great, but that will most likely need quite a bit of patching.

**Comment by sunflsks** [comment]
Yeah, a bit of work needs to be done before this is ready to merge, which is why I turned it into a draft.

**Comment by demhademha** [comment]
Would it also be possible to add [compiler-rt](https://compiler-rt.llvm.org)

**Comment by sunflsks** [comment]
I've reopened this as it mostly works. TAPI compile is currently broken for me, and cctools (which depends on it) it needs to be recompiled against LLVM 11. Besides that, it shouldn't have any major problems.

File: build_info/mono-jay.control [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Package: mono-jay
```
Comment (CRKatri) [inline build_info/mono-jay.control:1]: Don’t think this belongs here.

**Comment by demhademha** [comment]
Would you mind providing iPhoneOS-arm64 debs

**Comment by CRKatri** [comment]
clang-tools still needs to be debianized.

File: build_info/mono-jay.control (hunk 1 above)
Comment (sunflsks) [reply build_info/mono-jay.control:1]: Thanks for pointing that out

**Comment by CRKatri** [comment]
Also, symlink llvm-objcopy to objcopy.

**Comment by sunflsks** [comment]
> clang-tools still needs to be debianized.

Since each of the tools are trivial, and the total package size is fairly small (around 9 MB), I feel like this doesn't really need to be debianized. Providing a seperate package for each tool would be a pain for both the user and us, as many of these tools are used together.

**Comment by sunflsks** [comment]
> Would you mind providing iPhoneOS-arm64 debs

https://sudhip.com/files/other/llvm11.tar

**Comment by CRKatri** [comment]
> I've reopened this as it mostly works. TAPI compile is currently broken for me, and cctools (which depends on it) it needs to be recompiled against LLVM 11. Besides that, it shouldn't have any major problems.

So did you test cctools with llvm11? If you can’t get tapi to compile, just extract the debs into build_base to test it, and bump the dep and version ( you may also want to update cctools to remove the ldid wrapper removed a couple months ago)

**Comment by sunflsks** [comment]
> So did you test cctools with llvm11? 

Yes I did, and from my (very basic) testing, they work fine.

> If you can’t get tapi to compile, just extract the debs into build_base to test it, 

Yeah, this is exactly what I did.

> and bump the dep and version 

I've done this as well, I just need to push it.

> ( you may also want to update cctools to remove the ldid wrapper that got removed from upstream a couple months ago)

Could you elaborate, please?

**Comment by CRKatri** [comment]
> Could you elaborate, please?

https://github.com/tpoechtrager/cctools-port/commit/faa1f24cb7e31be132f98f503cc447c90ce2fd87

**Comment by CRKatri** [comment]
Also if you look into updating libtapi so we could get https://github.com/tpoechtrager/apple-libtapi/commit/664b8414f89612f2dfd35a9b679c345aa5389026, and adding arm64e to lldb.

**Comment by sunflsks** [comment]
I've added `tapi` (and I'm pretty sure fixed tapi compile) in #795

**Comment by CRKatri** [comment]
🤦

**Comment by sunflsks** [comment]
?

**Comment by CRKatri** [comment]
I forgot you already did tapi.

**Comment by sunflsks** [comment]
Oh haha

**Comment by demhademha** [comment]
```
$clang a.c -o a -fuse-ld=lld
ld64.lld: warning: ignoring unknown argument: -no_deduplicate
ld64.lld: warning: ignoring unknown argument: -platform_version
ld64.lld: warning: -sdk_version is required when emitting min version load command.  Setting sdk version to match provided min version
Cannot open ios: No such file or directory
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)

```

**Comment by CRKatri** [comment]
I don't believe that lld supports iOS.

**Comment by demhademha** [comment]
Would it be possible to prevent `lld` from being compiled for `iOS`
Also, could lld be made to support iOS (with a patch), or would that not be possible (due to the internals of lld)

**Comment by CRKatri** [comment]
lld’s Mach-O support is lack luster, the main point is lld in procursus is for cross compiling for elf targets.

**Comment by sunflsks** [comment]
Getting rid of ld64 is will always be out of the question, as it's an essential part of the Darwin toolchain. But yeah, like you said, lld still has it's uses (even if it's not functional on iOS), like inspecting the generated executable and cross compiling

**Comment by demhademha** [comment]
Would it be possible to add the clang man-pages

File: llvm.mk [hunk 2]
```diff
@@ -3,12 +3,12 @@ $(error Use the main Makefile)
...
-LLVM_VERSION   := 10.0.0
-LLVM_MAJOR_V   := 10
-SWIFT_VERSION  := 5.3.2
+LLVM_VERSION   := 11.1.0
+LLVM_MAJOR_V   := 11
+SWIFT_VERSION  := 5.4.1
 SWIFT_SUFFIX   := RELEASE
 DEB_SWIFT_V    ?= $(SWIFT_VERSION)~$(SWIFT_SUFFIX)
-DEB_LLVM_V     ?= $(LLVM_VERSION)~$(DEB_SWIFT_V)
+DEB_LLVM_V     ?= $(LLVM_VERSION)~$(DEB_SWIFT_V)-1
```
Comment (CRKatri) [inline llvm.mk:11]: You don't need this `-1`

File: llvm.mk [hunk 3]
```diff
@@ -184,14 +202,20 @@ llvm-package: llvm-stage
...
+	cp -a $(BUILD_STAGE)/llvm/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/llvm-$(LLVM_MAJOR_V)/lib/{libPollyISL.a,LLVMPolly.so,libPollyPPCG.a,libPolly.a} $(BUILD_DIST)/libllvm-polly$(LLVM_MAJOR_V)/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib
+	cp -a $(BUILD_STAGE)/llvm/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/llvm-$(LLVM_MAJOR_V)/lib/cmake/polly $(BUILD_DIST)/libllvm-polly$(LLVM_MAJOR_V)/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib
+
 	# llvm.mk Prep libclang-common-$(LLVM_MAJOR_V)-dev
 	mkdir -p $(BUILD_DIST)/libclang-common-$(LLVM_MAJOR_V)-dev/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/{,llvm-$(LLVM_MAJOR_V)/lib/}clang
-	cp -a $(BUILD_STAGE)/llvm/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/llvm-$(LLVM_MAJOR_V)/lib/clang/$(LLVM_VERSION) $(BUILD_DIST)/libclang-common-$(LLVM_MAJOR_V)-dev/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/llvm-$(LLVM_MAJOR_V)/lib/clang
+	cp -a $(BUILD_STAGE)/llvm/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/llvm-$(LLVM_MAJOR_V)/lib/clang/11.1.0 $(BUILD_DIST)/libclang-common-$(LLVM_MAJOR_V)-dev/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/llvm-$(LLVM_MAJOR_V)/lib/clang
```
Comment (CRKatri) [inline llvm.mk:213]: Use LLVM_VERSION here.

**Comment by CRKatri** [comment]
Also, I thing Apple clang was on 12? Or are the Swift releases still using 11?

File: build_info/clang-tools.control [hunk 4]
```diff
@@ -0,0 +1,10 @@
+Package: clang-tools
+Version: @DEB_LLVM_V@
+Architecture: @DEB_ARCH@
+Maintainer: @DEB_MAINTAINER@
+Depends: clang-tools-11
+Section: Development
+Priority: optional
+Homepage: https://www.llvm.org/
+Description: Extra clang tools
+Tag: purpose::console, role::developer
```
Comment (CRKatri) [inline build_info/clang-tools.control:10]: Have this provide clangd, clang-format, clang-tidy.

File: build_info/clang-tools-11.control [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Package: clang-tools-11
+Version: @DEB_LLVM_V@
+Architecture: @DEB_ARCH@
+Maintainer: @DEB_MAINTAINER@
+Depends: clang-11, python3, libllvm11, libclang-cpp11, libclang-common-11-dev
+Section: Development
+Priority: optional
+Homepage: https://www.llvm.org/
+Description: Extra clang tools
+Tag: purpose::console, role::developer
```
Comment (CRKatri) [inline build_info/clang-tools-11.control:10]: ditto

**Comment by sunflsks** [comment]
> Also, I thing Apple clang was on 12? Or are the Swift releases still using 11?

Apple clang's version string is 12.0.5, but the upstream LLVM version it uses is 11.1.0. For Xcode 13, they've switched to upstream LLVM 12

**Comment by Diatrus** [comment]
I'll update tapi and cctools in a separate commit. Thank you!

**Pull request merged**